Implement vsomeip-sys to wrap vsomeip C++ library

PLeVasseur opened the pull request:
Implements #11

Breaking this off as PR number 1/2 from #2

To Dos

- [x] Add CI

Opened an issue for adding unit and integration tests

devkelley commented on .github/workflows/lint-and-test.yaml:
```diff
@@ -28,16 +40,171 @@ concurrency:
+          echo "VSOMEIP install path from previous job: ${{ env.VSOMEIP_INSTALL_PATH }}"
+          ls -l ${{ env.VSOMEIP_INSTALL_PATH }}
+
+# TODO: Temporarily commenting this proven bit out to get the download & compilation
```
Can be removed now?

PLeVasseur replied:
Again... haha, feel free to review everything but the CI :)

PLeVasseur replied:
Fixed

devkelley commented on .github/workflows/lint-and-test.yaml:
```diff
@@ -28,16 +40,171 @@ concurrency:
+          path: ${{ env.VSOMEIP_INSTALL_PATH }}
+          key: ${{ runner.os }}-vsomeip-${{ hashFiles(env.VSOMEIP_SOURCE_PATH_WITH_WILDCARD) }}
+
+  foo:
```
maybe a more descriptive job name?

PLeVasseur replied:
Hah, sorry. I'm still in the midst of updating the CI. Feel free to review everything but that :)

devkelley replied:
fair enough :)

PLeVasseur replied:
foo removed :)

devkelley commented on Cargo.toml:
```diff
@@ -25,10 +25,10 @@ exclude = [
 readme = "README.md"
 repository = "https://github.com/eclipse-uprotocol/up-client-vsomeip-rust"
-rust-version = "1.72"
+rust-version = "1.74"
```
why 1.74 rather than 1.76? To keep parity with the up-rust library?

PLeVasseur replied:
up-transport-zenoh-rust uses 1.74

devkelley replied:
Got it, and what was the reasoning behind removing the toolchain file?

PLeVasseur replied:
cargo tarpaulin was failing at 1.74 and I need to work through why. Until then, I just removed the toolchain file :upside_down_face:

devkelley replied:
Ah got it! FWIW, I had a similar issue with tarpaulin and it resolved itself once I upgraded to 1.76. Has something to do with tarpaulin using a feature not available until that version

PLeVasseur replied:
Gotcha. Zenoh will upgrade to 1.75 for their 1.0.0 release, so I was hoping it'd just resolve then. But maybe not hah

PLeVasseur commented:
Okay @devkelley & all: the CI is now updated to work!

We can see it run a single unit test in vsomeip-sys I had written in the past :)

https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust/actions/runs/9786728167/job/27022105149

PLeVasseur commented:
Taking off of draft status now that CI works! :slightly_smiling_face:

PLeVasseur commented on .github/workflows/lint-and-test.yaml:
```diff
@@ -13,6 +13,18 @@
+  VSOMEIP_SOURCE_PATH: vsomeip-src
+  VSOMEIP_INSTALL_PATH: vsomeip-install
+
+  # TODO: Need to get a cached version of vsomeip compiled we can refer to here
```
We do have a cached version and the variable used for this is `VSOMEIP_INSTALL_PATH` from above. Should delete this TODO.

PLeVasseur commented on .github/workflows/lint-and-test.yaml:
```diff
@@ -13,6 +13,18 @@
+
+  # TODO: Need to get a cached version of vsomeip compiled we can refer to here
+  VSOMEIP_LIB_PATH: /does/not/yet/exist
+  # TODO: Need to get a C++ build environment setup with C++ libs in known locations
```
This is also done, should remove TODO

stevenhartley commented on vsomeip-sys/build.rs:
```diff
@@ -0,0 +1,264 @@
+            "pub use bindgen::root::std_chrono_duration_long_AutocxxConcrete",
+            "#[allow(unused_imports)] pub use bindgen::root::std_chrono_duration_long_AutocxxConcrete"
+        );
+        line = line.replace(
```
Do we not want to create a patch and apply it like what Ivan was doing for the vsomeip helloworld example? Actually looking over the code again it looks like you are overriding the rust generated code is that correct? Why?

PLeVasseur replied:
Yes, overriding the generated Rust code in a few spots to:
1. eliminate warnings
2. add the ability to output enum fields

stevenhartley commented on vsomeip-sys/src/lib.rs:
```diff
@@ -0,0 +1,144 @@
+    #include "application_wrapper.h"
+    #include "message_wrapper.h"
+    #include "payload_wrapper.h"
+    safety!(unsafe) // see details of unsafety policies described in the 'safety' section of the book
```
unsafe? Why?

PLeVasseur replied:
We're wrapping C++ code, which seems to mean unsafe by default.

stevenhartley reviewed (commented):
95% of the code is over my head and hard to follow. I hope @sophokles73 and @devkelley can help cover there. I wonder if we should put vsomeip-sys Inside of COVESA like we did for vsomeip-python project, what do you think?

PLeVasseur commented:
> 95% of the code is over my head and hard to follow. I hope @sophokles73 and @devkelley can help cover there. I wonder if we should put vsomeip-sys Inside of COVESA like we did for vsomeip-python project, what do you think?

From what I saw of vsomeip-python it looked like a more complete wrapper. vsomeip-sys is, for the moment, just enough to enable building up-transport-vsomeip-rust.

If we have the time / inclination to cover a broader area of vsomeip and a larger maintenance load, it's possible to upstream.

sophokles73 commented on vsomeip-sys/README.md:
````diff
@@ -0,0 +1,57 @@
+
+This is required because of symlinking that's performed by the vsomeip library install (which is typical).
+
+## Supplying your own vsomeip includes
+
+You may modify the build command above to include setting the environment variable:
+
+```
+VSOMEIP_INCLUDE_PATH=<path/to/vsomeip/include>
+```
+
+`VSOMEIP_INCLUDE_PATH` must point to the folder which has the `vsomeip` folder within it containing the vsomeip includes.
````
My understanding is that we _always_ require the user to have the vsomeip shared library installed, i.e. we do not create it as part of the vsomeip-sys build.
If that is the case, then I would suggest to also require the user to _always_ provide the (matching) header files. Otherwise we might run into situations where the user provides the lib in version A and we download the source code in version B. This doesn't seem to pose any additional effort for the user because he needs to build vsomeip from source anyway, as there is no binary distribution and as part of the build/install process the vsomeip headers will also be installed, right?

PLeVasseur replied:
Added the ability to do this in the latest commit. Looking for feedback

sophokles73 commented on vsomeip-sys/build.rs:
```diff
@@ -0,0 +1,264 @@
+        if let Ok(user_supplied_vsomeip_include_path) = user_supplied_vsomeip_include_path {
+            PathBuf::from(user_supplied_vsomeip_include_path)
+        } else {
+            let vsomeip_archive_dest = Path::new(&out_dir).join("vsomeip").join("vsomeip.tar.gz");
+            let vsomeip_archive_url =
+                format!("{VSOMEIP_TAGGED_RELEASE_BASE}{VSOMEIP_VERSION_ARCHIVE}");
+            let vsomeip_source_folder = Path::new(&out_dir).join("vsomeip").join("vsomeip-src");
+            download_and_write_file(&vsomeip_archive_url, &vsomeip_archive_dest)
+                .expect("Unable to download released archive");
+            decompress::decompress(
+                vsomeip_archive_dest,
+                vsomeip_source_folder.clone(),
+                &ExtractOptsBuilder::default().strip(1).build().unwrap(),
+            )
+            .expect("Unable to extract tar.gz");
+            vsomeip_source_folder.join("interface")
```
FMPOV we should rely on the user always providing the path

PLeVasseur replied:
Yeah we chatted about this. I'll probably have to make this change for the reasons you outlined above.

PLeVasseur replied:
Added the ability to compile the downloaded vsomeip library for the user in the latest commit. Looking for feedback :slightly_smiling_face:

sophokles73 commented on vsomeip-sys/build.rs:
```diff
@@ -0,0 +1,264 @@
+    let runtime_wrapper_dir = project_root.join("src/glue/include"); // Update the path as necessary
+
+    // Somewhat useful debugging
+    // println!("cargo:warning=# CARGO_MANIFEST_DIR : {}", project_root.display());
+    // println!("cargo:warning=# OUT_DIR            : {}", out_path.display());
+    // println!("cargo:warning=# vsomeip_interface  : {}", interface_path.display());
+    // println!("cargo:warning=# runtime_wrapper    : {}", runtime_wrapper_dir.display());
```
if this is really useful then why is it commented out?

PLeVasseur replied:
Good point. I need to find a better way that doesn't show up as a yellow, scary warning :sweat_smile:

PLeVasseur replied:
I couldn't find a way to make this a non-scary yellow warning, but I find it to be helpful regardless

sophokles73 commented on vsomeip-sys/build.rs:
```diff
@@ -0,0 +1,264 @@
+        .compile("cxx-portion");
+    println!("cargo:rerun-if-changed=src/cxx_bridge.rs");
+
+    // we rewrite the autocxx generated code to suppress the cargo warning about unused imports
```
why do we need to suppress these warnings? Do they break the build?

PLeVasseur replied:
They break the build since we fail on warnings emitted from the generated Rust code

PLeVasseur replied:
Updated to clarify what's going on by breaking apart

sophokles73 commented on vsomeip-sys/build.rs:
```diff
@@ -0,0 +1,264 @@
+        );
+        line = line.replace(
+            "pub unsafe fn set_data(self: Pin<&mut payload>, _data: *const u8, _length: u32);",
+            "pub(crate) unsafe fn set_data(self: Pin<&mut payload>, _data: *const u8, _length: u32);"
```
why do you need to reduce visibility here?

PLeVasseur replied:
Broke these apart to clarify

sophokles73 commented on vsomeip-sys/build.rs:
```diff
@@ -0,0 +1,264 @@
+// provided by `destination` parameter.
+fn download_and_write_file(url: &str, dest_path: &PathBuf) -> Result<(), Box<dyn Error>> {
+    let client = Client::builder()
+        .timeout(Duration::from_secs(120)) // Set a timeout of 60 seconds
```
looks like a timeout of 120 secs to me, anyway, the comment seems superfluous to me ;-)

PLeVasseur replied:
Resolved :laughing:

sophokles73 commented on vsomeip-sys/src/lib.rs:
```diff
@@ -0,0 +1,144 @@
+}
+
+pub mod extern_callback_wrappers;
+pub mod safe_glue;
```
having a public API in a module called `glue` and `safe_glue` feels a little odd. Do you really need to expose these modules in their entirety? Maybe you can simple reexport some types of rename the modules?

PLeVasseur replied:
Generally speaking... I should probably revisit the module setup.

I agree with you. I may open an issue for this instead of addressing on this PR since it'll take a bit of rearrangement.

sophokles73 commented on vsomeip-sys/src/safe_glue.rs:
```diff
@@ -0,0 +1,455 @@
+/// # TODO
+///
+/// Add some runtime safety checks on the pointer
+pub fn get_pinned_runtime(wrapper: &RuntimeWrapper) -> Pin<&mut runtime> {
```
can these explicit functions maybe replaced with some clever implementation of `Deref` instead?

PLeVasseur replied:
I guess this is not allowed since the borrow checker will not allow for pinned mutable borrows to be returned from Deref.

I spent some time on this today, but didn't see a way around the issue.

Darn. I'm going to call this one unsolvable for the moment :(

PLeVasseur replied:
I updated this to clarify and hopefully simplify the API.

sophokles73 reviewed (commented):
I have deliberately skipped reviewing any of the c++ files since I have no clue about c++ ;-) So this is as far as I can support ...

PLeVasseur commented on up-transport-vsomeip/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+################################################################################
+# Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Done

PLeVasseur commented on vsomeip-sys/Cargo.toml:
```diff
@@ -0,0 +1,34 @@
+cxx-build = "1.0.122"
+decompress = { version = "0.6.0" }
+miette = { version = "5", features = ["fancy"] }
+reqwest = { version = "0.12", features = ["blocking"] }
```
Done

sophokles73 commented on vsomeip-sys/build.rs:
```diff
@@ -134,81 +143,72 @@ fn main() -> miette::Result<()> {
+fn fix_unused_imports(line: String) -> String {
+    let mut fixed_line = line.replace(
+        "pub use bindgen::root::std_chrono_duration_long_AutocxxConcrete",
+        "#[allow(unused_imports)] pub use bindgen::root::std_chrono_duration_long_AutocxxConcrete",
```
Am I right in assuming that the generated code itself does not need the type but that the vsomeip transport will need it?

PLeVasseur replied:
I think we don't use it in the transport. Would you recommend removal entirely then?

sophokles73 replied:
Unless you want to create a generic vsomeip-sys crate that is independent of vsomeip-utransport ...

PLeVasseur replied:
No thanks, too much work atm :melting_face: 

Will remove

PLeVasseur replied:
Done

sophokles73 commented on vsomeip-sys/build.rs:
```diff
@@ -134,81 +143,72 @@ fn main() -> miette::Result<()> {
-        //     "# [repr (u8)] # [derive (Clone , Hash , PartialEq , Eq)] pub enum message_type_e",
-        //     "# [repr (u8)] # [derive (Clone , Hash , PartialEq , Eq, Debug)] pub enum message_type_e"
-        // );
+        line = fix_unused_imports(line);
+        line = fix_unsafe_fn_unused(line);
+        line = fix_doc_build(line);
+        line = add_enum_debug(line);
```
this is much better :+1:

sophokles73 reviewed (commented):
I like the idea of being able to build the vsomeip library as part of the transport's build. Have you taken a look at other *-sys crates and how they do this? Maybe https://github.com/eclipse/paho.mqtt.rust/tree/master/paho-mqtt-sys can provide some additional inspiration ...

PLeVasseur commented on vsomeip-sys/build.rs:
```diff
@@ -60,10 +60,19 @@ fn main() -> miette::Result<()> {
-    // println!("cargo:warning=# OUT_DIR            : {}", out_path.display());
-    // println!("cargo:warning=# vsomeip_interface  : {}", interface_path.display());
-    // println!("cargo:warning=# runtime_wrapper    : {}", runtime_wrapper_dir.display());
+    println!("cargo:warning=# vsomeip_lib_path  : {}", vsomeip_lib_path);
```
Lemme look into this

PLeVasseur replied:
Jammed pack full of lots of other debug stuff if we do this, but it's still more in the spirit of being a debug message, so I'll do it.

PLeVasseur replied:
Done

PLeVasseur commented on .github/workflows/lint-and-test.yaml:
```diff
@@ -28,16 +40,132 @@ concurrency:
       cancel-in-progress: true
 
 jobs:
+  download_and_decompress:
```
I'm going to open an issue to look into this more, but mark resolved for now to keep moving.

PLeVasseur replied:
Okay, I now test for both with the bundled feature flag turned on and off.

PLeVasseur commented:
> I like the idea of being able to build the vsomeip library as part of the transport's build. Have you taken a look at other *-sys crates and how they do this? Maybe https://github.com/eclipse/paho.mqtt.rust/tree/master/paho-mqtt-sys can provide some additional inspiration ...

Seems like the answer is "use a git submodule", much like you did for up-rust to use the up-spec protos.

I'll look into this.

sophokles73 commented on .github/workflows/lint-and-test.yaml:
```diff
@@ -63,21 +111,21 @@ jobs:
         run: |
           # enable nightly features so that we can also include Doctests
-          RUSTC_BOOTSTRAP=1 cargo tarpaulin -o xml -o lcov -o html --doc --tests
+          RUSTC_BOOTSTRAP=1 cargo tarpaulin -o xml -o lcov -o html --doc --tests -- --test-threads 1
```
on my local Ubuntu 22.04 machine I need to explicitly set LD_LIBRARY_PATH to also include `${VSOMEIP_INSTALL_PATH}/lib` for the tests to run successfully ...

PLeVasseur replied:
Added this

sophokles73 commented on Cargo.toml:
```diff
@@ -25,22 +31,22 @@ exclude = [
 serde = { version = "1.0", features = ["derive"] }
 serde_json = { version = "1.0" }
-up-rust = { git = "https://github.com/eclipse-uprotocol/up-rust", version = "0.1.5" }
+up-rust = { git = "https://github.com/eclipse-uprotocol/up-rust", rev = "3a50104421a801d52e1d9c68979db54c013ce43d" }
```
is there a reason why you switched back to the commit instead of e.g. using version `0.1`?

PLeVasseur replied:
Didn't switch. I mentioned on the other PR that I have a separate PR updating to 0.1 over [here](https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust/pull/13).

sophokles73 reviewed (commented):
Being able to build vsomeip-sys using the `bundled` feature is great :+1:

PLeVasseur commented on .github/workflows/lint-and-test.yaml:
```diff
@@ -28,16 +33,38 @@ concurrency:
       - uses: actions/checkout@v4
+
+      - name: Install C++ dependencies
+        run: sudo apt-get install -y build-essential cmake libboost-all-dev
```
Added this

PLeVasseur commented on vsomeip-sys/README.md:
````diff
@@ -0,0 +1,82 @@
+If you are running a Linux-based OS on x86_64, it's possible these look like:
+
+```
+GENERIC_CPP_STDLIB_PATH = /usr/include/c++/11
````
Fixed the formatting

PLeVasseur commented:
> Being able to build vsomeip-sys using the `bundled` feature is great 👍

Hey @sophokles73 -- can you take another look?

I re-added build steps in which we don't use the `bundled` feature to make sure we support that use case.

Also addressed the comments you had left.

Thx!

sophokles73 reviewed (approved):
LGTM
